File: packages/core/src/ssr/htmlBuilder.ts
export type HTMLAttributeValue = string | number | boolean | null | undefined;

export class HTMLBuilder {
  private parts: string[] = [];

  openTag(tag: string, attrs?: Record<string, HTMLAttributeValue>): this {
    let html = `<${tag}`;
    if (attrs) {
      for (const [key, value] of Object.entries(attrs)) {
        if (value === true) {
          html += ` ${key}`;
        } else if (value !== false && value !== null && value !== undefined) {
          html += ` ${key}="${escapeHtml(String(value))}"`;
        }
      }
    }
    html += ">";
    this.parts.push(html);
    return this;
  }

  tag(tag: string, attrs: Record<string, HTMLAttributeValue> | undefined, content: string): this {
    let html = `<${tag}`;
    if (attrs) {
      for (const [key, value] of Object.entries(attrs)) {
        if (value === true) {
          html += ` ${key}`;
        } else if (value !== false && value !== null && value !== undefined) {
          html += ` ${key}="${escapeHtml(String(value))}"`;
        }
      }
    }
    html += ">";
    html += escapeHtml(content);
    html += `</${tag}>`;
    this.parts.push(html);
    return this;
  }

  closeTag(tag: string): this {
    this.parts.push(`</${tag}>`);
    return this;
  }

  text(content: string): this {
    this.parts.push(escapeHtml(content));
    return this;
  }

  html(content: string): this {
    this.parts.push(content);
    return this;
  }

  build(): string {
    return this.parts.join("");
  }
}

export function escapeHtml(text: string): string {
  return text
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/\"/g, "&quot;")
    .replace(/'/g, "&#39;");
}

File: ssr/extable-go/htmlbuilder.go
package extable

import "strings"

type htmlBuilder struct {
	sb strings.Builder
}

func (b *htmlBuilder) openTag(tag string, attrs ...string) {
	b.sb.WriteString("<")
	b.sb.WriteString(tag)
	for i := 0; i+1 < len(attrs); i += 2 {
		key := attrs[i]
		value := attrs[i+1]
		if key == "" {
			continue
		}
		b.sb.WriteString(" ")
		b.sb.WriteString(key)
		b.sb.WriteString("=\"")
		b.sb.WriteString(escapeHTML(value))
		b.sb.WriteString("\"")
	}
	b.sb.WriteString(">")
}

func (b *htmlBuilder) closeTag(tag string) {
	b.sb.WriteString("</")
	b.sb.WriteString(tag)
	b.sb.WriteString(">")
}

func (b *htmlBuilder) text(text string) {
	b.sb.WriteString(escapeHTML(text))
}

func (b *htmlBuilder) raw(html string) {
	b.sb.WriteString(html)
}

func (b *htmlBuilder) string() string {
	return b.sb.String()
}
